### ISSUE TITLE
YaRN的疑惑

### ISSUE THREAD

**Issue by github-luffy** [jingyaogong/minimind#602]
感谢大佬的开源，阅读源码有个疑惑：对比了两个版本的YaRN代码实现，为什么high最后的取值不一样，一个是min(, dim // 2 - 1)，一个是min(high, dim-1)？我试过一些数据跑了下两份代码，最后high的取值确实会不一样，这两种情况都可以的吗？

<img width="823" height="140" alt="Image" src="https://github.com/user-attachments/assets/8f8a05ef-3d3e-4a82-b297-3b99cc77dc36" />

<img width="754" height="153" alt="Image" src="https://github.com/user-attachments/assets/645d17b0-c432-4243-b551-5bef14f47181" />

本项目代码如下：

def precompute_freqs_cis(dim: int, end: int = int(32 * 1024), rope_base: float = 1e6, rope_scaling: Optional[dict] = None):
      freqs, attn_factor = 1.0 / (rope_base ** (torch.arange(0, dim, 2)[: (dim // 2)].float() / dim)), 1.0
      if rope_scaling is not None:
          orig_max, factor, beta_fast, beta_slow, attn_factor = (
              rope_scaling.get("original_max_position_embeddings", 2048), rope_scaling.get("factor", 16),
              rope_scaling.get("beta_fast", 32.0), rope_scaling.get("beta_slow", 1.0), rope_scaling.get("attention_factor", 1.0)
          )
          if end / orig_max > 1.0:
              # YaRN: f'(i) = f(i)((1-γ) + γ/s), where γ∈[0,1] is linear ramp
              inv_dim = lambda b: (dim * math.log(orig_max / (b * 2 * math.pi))) / (2 * math.log(rope_base))
              low, high = max(math.floor(inv_dim(beta_fast)), 0), min(math.ceil(inv_dim(beta_slow)), dim // 2 - 1)
              ramp = torch.clamp((torch.arange(dim // 2, device=freqs.device).float() - low) / max(high - low, 0.001), 0, 1)
              freqs = freqs * (1 - ramp + ramp / factor)
    
      t = torch.arange(end, device=freqs.device)
      freqs = torch.outer(t, freqs).float()
      freqs_cos = torch.cat([torch.cos(freqs), torch.cos(freqs)], dim=-1) * attn_factor
      freqs_sin = torch.cat([torch.sin(freqs), torch.sin(freqs)], dim=-1) * attn_factor
      return freqs_cos, freqs_sin

DeepSeek版本代码如下：
def precompute_freqs_cis(args: ModelArgs) -> torch.Tensor:
    """
    Precomputes frequency-based complex exponential values for rotary positional embeddings.

    Args:
        args (ModelArgs): Model arguments containing positional embedding parameters.

    Returns:
        torch.Tensor: Precomputed complex exponential values for positional embeddings.
    """
    dim = args.qk_rope_head_dim
    seqlen = args.max_seq_len
    beta_fast = args.beta_fast
    beta_slow = args.beta_slow
    base = args.rope_theta
    factor = args.rope_factor

    def find_correction_dim(num_rotations, dim, base, max_seq_len):
        """
        Computes the correction dimension for a given number of rotations in the rotary positional embedding.

        Args:
            num_rotations (float): Number of rotations to compute the correction for.
            dim (int): Dimensionality of the embedding space.
            base (float): Base value for the exponential computation.
            max_seq_len (int): Maximum sequence length.

        Returns:
            float: The correction dimension based on the input parameters.
        """
        return dim * math.log(max_seq_len / (num_rotations * 2 * math.pi)) / (2 * math.log(base))

    def find_correction_range(low_rot, high_rot, dim, base, max_seq_len):
        """
        Computes the range of correction dimensions for rotary positional embeddings.

        Args:
            low_rot (float): Lower bound for the number of rotations.
            high_rot (float): Upper bound for the number of rotations.
            dim (int): Dimensionality of the embedding space.
            base (float): Base value for the exponential computation.
            max_seq_len (int): Maximum sequence length.

        Returns:
            Tuple[int, int]: The range of correction dimensions (low, high), clamped to valid indices.
        """
        low = math.floor(find_correction_dim(low_rot, dim, base, max_seq_len))
        high = math.ceil(find_correction_dim(high_rot, dim, base, max_seq_len))
        return max(low, 0), min(high, dim-1)

    def linear_ramp_factor(min, max, dim):
        """
        Computes a linear ramp function used to smooth values between a minimum and maximum range.

        Args:
            min (float): Minimum value for the ramp function.
            max (float): Maximum value for the ramp function.
            dim (int): Dimensionality of the ramp tensor.

        Returns:
            torch.Tensor: A tensor of shape (dim,) with values linearly interpolated between 0 and 1,
                clamped to the range [0, 1].
        """
        if min == max:
            max += 0.001
        linear_func = (torch.arange(dim, dtype=torch.float32) - min) / (max - min)
        ramp_func = torch.clamp(linear_func, 0, 1)
        return ramp_func

    freqs = 1.0 / (base ** (torch.arange(0, dim, 2, dtype=torch.float32) / dim))
    if seqlen > args.original_seq_len:
        low, high = find_correction_range(beta_fast, beta_slow, dim, base, args.original_seq_len)
        smooth = 1 - linear_ramp_factor(low, high, dim // 2)
        freqs = freqs / factor * (1 - smooth) + freqs * smooth

    t = torch.arange(seqlen)
    freqs = torch.outer(t, freqs)
    freqs_cis = torch.polar(torch.ones_like(freqs), freqs)
    return freqs_cis

希望大佬有空回下，解答下我的困惑！

**Comment by jingyaogong** [comment]
😄感谢认可，结论先行：

没错但可能没必要；
从上面打印的截图目测应该是「rope_base太小」或「ori_len太大」的原因（因为dim似乎=16？high取到了9），可以调整再看；

---

emm数学上限确实是 `dim - 1`，不过 `dim // 2 - 1` 现实完全没区别

因为 `dim = 2 × length`，频率索引范围是 `[0, length-1]`，分成三段：
- `[0, low)`：高频不动区域
- `[low, high]`：插值过渡区域  
- `(high, length-1]`：低频纯缩放区域

直觉上 `high ≤ length - 1 = dim//2 - 1`

但存在 `high > length-1` 的情况，说白了此时低频区不存在，`high` 算出来突破 `dim//2 - 1` 的限制

具体什么情况会超出？

```python
high = min(ceil((dim × log(orig_max / 2π)) / (2 × log(rope_base))), 上限)
```
也就是当 `orig_max > 2π × rope_base ≈ 6.28 × rope_base` 时，理论 `high` 值会超出 `dim//2 - 1` 上限

但是

MiniMind：
```bash
dim=64, orig_max=2048, rope_base=1e6
inv_dim(1) = 13.40
high = min(14, 31) = 14 # dim//2-1=31够用，14远未超出，没必要用dim-1=63
```

DeepSeek V3x：
```bash
dim=64, orig_max=4096, rope_base=10000
inv_dim(1) = 22.51
high = min(23, 31) = 23  # dim//2-1=31够用，23未超出，没必要用63
```

极端情况（`orig_max=65000, rope_base=10000`）会超出：
```bash
inv_dim(1) = 32.12 > 31 = dim//2-1  # 确实超出，得用63了
```
意味着 `rope_base` 远远跟不上原始上下文长度时才会发生这种情况（此时兜底是有意义的，但即使此时截断到dim//2-1也无非是让过渡“陡峭”一些，没大问题）

之所以没有实际的必要，是因为实际的 LLM 不会这样。具体 `rope_base` 从 100 到 10000 的对比，小 base 势必导致高频振荡问题位置分辨率直接错乱（现在主流模型base都开的挺大）：

![rope_base对比](https://github.com/user-attachments/assets/1099d529-48ce-4f56-878a-0a103a71cdeb)


可以参考下现在主流模型 `rope_base` 普遍采用 **1e6 ～ 5e6**：

https://huggingface.co/zai-org/GLM-4.7/blob/main/config.json
https://huggingface.co/MiniMaxAI/MiniMax-M2/blob/main/config.json
https://huggingface.co/Qwen/Qwen3-4B-Instruct-2507/blob/main/config.json
...

如果是 `rope_base=1e6`，发生以上临界是 `orig_max = 6.28 × 1e6 ≈ 628 万`，6280k原生上下文远远远远远超任何模型实际训练长度。

另外，ramp 函数与 low/high 的插值关系（前两个是即使改dim-1也没有区别，第三个是极端情况，会产生差异，也就是前面说的没有低频纯插，和你图中一致的情况）：

![ramp插值图](https://github.com/user-attachments/assets/e5219086-a0e1-4c2a-a527-c8424f419bd9)

**Comment by github-luffy** [comment]
> 😄感谢认可，结论先行：
> 
> 没错但可能没必要； 从上面打印的截图目测应该是「rope_base太小」或「ori_len太大」的原因（因为dim似乎=16？high取到了9），可以调整再看；
> 
> emm数学上限确实是 `dim - 1`，不过 `dim // 2 - 1` 现实完全没区别
> 
> 因为 `dim = 2 × length`，频率索引范围是 `[0, length-1]`，分成三段：
> 
> * `[0, low)`：高频不动区域
> * `[low, high]`：插值过渡区域
> * `(high, length-1]`：低频纯缩放区域
> 
> 直觉上 `high ≤ length - 1 = dim//2 - 1`
> 
> 但存在 `high > length-1` 的情况，说白了此时低频区不存在，`high` 算出来突破 `dim//2 - 1` 的限制
> 
> 具体什么情况会超出？
> 
> high = min(ceil((dim × log(orig_max / 2π)) / (2 × log(rope_base))), 上限)
> 也就是当 `orig_max > 2π × rope_base ≈ 6.28 × rope_base` 时，理论 `high` 值会超出 `dim//2 - 1` 上限
> 
> 但是
> 
> MiniMind：
> 
> dim=64, orig_max=2048, rope_base=1e6
> inv_dim(1) = 13.40
> high = min(14, 31) = 14 # dim//2-1=31够用，14远未超出，没必要用dim-1=63
> DeepSeek V3x：
> 
> dim=64, orig_max=4096, rope_base=10000
> inv_dim(1) = 22.51
> high = min(23, 31) = 23  # dim//2-1=31够用，23未超出，没必要用63
> 极端情况（`orig_max=65000, rope_base=10000`）会超出：
> 
> inv_dim(1) = 32.12 > 31 = dim//2-1  # 确实超出，得用63了
> 意味着 `rope_base` 远远跟不上原始上下文长度时才会发生这种情况（此时兜底是有意义的，但即使此时截断到dim//2-1也无非是让过渡“陡峭”一些，没大问题）
> 
> 之所以没有实际的必要，是因为实际的 LLM 不会这样。具体 `rope_base` 从 100 到 10000 的对比，小 base 势必导致高频振荡问题位置分辨率直接错乱（现在主流模型base都开的挺大）：
> 
> ![rope_base对比](https://github.com/user-attachments/assets/1099d529-48ce-4f56-878a-0a103a71cdeb)
> 
> 可以参考下现在主流模型 `rope_base` 普遍采用 **1e6 ～ 5e6**：
> 
> https://huggingface.co/zai-org/GLM-4.7/blob/main/config.json https://huggingface.co/MiniMaxAI/MiniMax-M2/blob/main/config.json https://huggingface.co/Qwen/Qwen3-4B-Instruct-2507/blob/main/config.json ...
> 
> 如果是 `rope_base=1e6`，发生以上临界是 `orig_max = 6.28 × 1e6 ≈ 628 万`，6280k原生上下文远远远远远超任何模型实际训练长度。
> 
> 另外，ramp 函数与 low/high 的插值关系（前两个是即使改dim-1也没有区别，第三个是极端情况，会产生差异，也就是前面说的没有低频纯插，和你图中一致的情况）：
> 
> ![ramp插值图](https://github.com/user-attachments/assets/e5219086-a0e1-4c2a-a527-c8424f419bd9)

感谢您的回复，我再好好看看代码。
